### PR TITLE
add `SSLSocket.open`

### PR DESCRIPTION
**Pull request by thekuwayama**
Hi!

This PR is related to #225 .
* add `SSLSocket.open` as alias of `SSLSocket.new`

### CODE REVIEW

File: test/test_ssl.rb [hunk 1]
```diff
@@ -1586,6 +1586,17 @@ def test_fileno
     sock2.close
   end
 
+  def test_ssl_socket_new_alias
+    mn = OpenSSL::SSL::SSLSocket.method(:new)
+    mo = OpenSSL::SSL::SSLSocket.method(:open)
+
+    if mn.inspect == "#<Method: Class#new>"
```
Comment (thekuwayama) [inline test/test_ssl.rb]: Review this test. I wonder how to test this.

**Comment by thekuwayama** [comment]
Oh, the other test failed (and I can't restart).

**Comment by ioquatix** [comment]
Should it be an alias or do we need to change the functionality? Because I assume we should at least try to match the same as `TCPSocket`. Do you mind checking it? What are the semantics, the arguments, etc.

**Comment by ioquatix** [comment]
Can you rebase on master, tests should be fixed.

**Comment by thekuwayama** [comment]
I think that we do not need to change the function.

**Comment by ioquatix** [comment]
I am really happy with the direction this is going in, but I'm not sure about the spec.

I feel like we might need to add some documentation too, to make it clear about how it should be used. Even thought it's an alias, we don't want people using it in the same way as `SSLSocket.new`...

In an ideal world, `SSLSocket.new` takes in a file descriptor, and `SSLSocket.open` can take a hostname and/or address/port.

Since we have an opportunity now to make that the normal behaviour, I wonder if we should make things a bit tighter.

**Comment by thekuwayama** [comment]
> In an ideal world, SSLSocket.new takes in a file descriptor, and SSLSocket.open can take a hostname and/or address/port.

OK. Sounds good to me, I updated API concerning `TCPSocket.open`.

```
open(remote_host, remote_port, context=nil, local_host=nil, local_port=nil)
```

About the arguments, I think as follows:

- `context`, `local_host` and `local_port` are not required arguments.
    - with reference to `TCPSocket.open` and `SSLSocket.new`.
- `context` is 3rd argument, not 5th
    - `context` is more frequently used than `local_host` and `local_port`.

**Comment by ioquatix** [comment]
What do you think about making `context:` a keyword argument, so that the method signature is otherwise identical?

**Comment by thekuwayama** [comment]
Oh, that looks good idea!
For application layer such as `HTTP.rb`, It's better to provide a method signature that is identical to `TCPSocket.open`.

File: test/test_ssl.rb (hunk 1 above)
Comment (thekuwayama) [reply test/test_ssl.rb]: update method signature

**Comment by ioquatix** [comment]
Another interesting thing to consider is some useful default for context if it's nil.... maybe we can make it very strict.

**Comment by ioquatix** [comment]
I am very happy with this, well done!

**Comment by thekuwayama** [comment]
@ioquatix Thank you for reviewing!

**Pull request merged**